### PR TITLE
[inductor] parallel compile: Create new pipes for subproc communication

### LINKED ISSUE #131070: Subproc exception with torch complie for Torch 2.4.0 and Nightly
### 🐛 Describe the bug

Getting subproc exception with torch compile when onnxruntime-training and deepspeed packages are installed. This occurs for me both with Torch 2.4.0 and the nightly wheels.

Steps to reproduce:
```
docker run -it --gpus all --ipc host nvcr.io/nvidia/pytorch:24.06-py3
# Inside docker image
pip uninstall torch pytorch-quantization pytorch-triton torch-tensorrt torchvision -y
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
pip install onnxruntime-training
pip install deepspeed
python
```
```python
# Inside python run this:
import torch
def foo(x, y):
  a = torch.sin(x)
  b = torch.cos(y)
  return a + b

opt_foo = torch.compile(foo)

```

This is the error which I get:
```
>>> opt_foo = torch.compile(foo)
[2024-07-18 22:47:08,296] [INFO] [real_accelerator.py:203:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/usr/local/lib/python3.10/dist-packages/onnxruntime/capi/onnxruntime_validation.py:114: UserWarning: WARNING: failed to get cudart_version from onnxruntime build info.
  warnings.warn("WARNING: failed to get cudart_version from onnxruntime build info.")
>>> E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183] failure in SubprocPool._read_thread
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183] Traceback (most recent call last):
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183]   File "/usr/local/lib/python3.10/dist-packages/torch/_inductor/compile_worker/subproc_pool.py", line 161, in _read_thread
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183]     job_id, data = _recv_msg(self.read_pipe)
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183]   File "/usr/local/lib/python3.10/dist-packages/torch/_inductor/compile_worker/subproc_pool.py", line 61, in _recv_msg
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183]     data = read_pipe.read(length) if length > 0 else b""
E0718 22:47:11.256000 657 torch/_inductor/compile_worker/subproc_pool.py:183] MemoryError
/usr/local/lib/python3.10/dist-packages/onnxruntime/capi/onnxruntime_validation.py:114: UserWarning: WARNING: failed to get cudart_version from onnxruntime build info.
  warnings.warn("WARNING: failed to get cudart_version from onnxruntime build info.")
```

### Versions

PyTorch version: 2.5.0.dev20240718+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.29.3
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.160.1-1.cm2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.40
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 550.54.15
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.1.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7V12 64-Core Processor
CPU family:                           23
Model:                                49
Thread(s) per core:                   1
Core(s) per socket:                   48
Socket(s):                            2
Stepping:                             0
BogoMIPS:                             4890.87
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru arat umip rdpid
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            3 MiB (96 instances)
L1i cache:                            3 MiB (96 instances)
L2 cache:                             48 MiB (96 instances)
L3 cache:                             384 MiB (24 instances)
NUMA node(s):                         4
NUMA node0 CPU(s):                    0-23
NUMA node1 CPU(s):                    24-47
NUMA node2 CPU(s):                    48-71
NUMA node3 CPU(s):                    72-95
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; untrained return thunk; SMT disabled
Vulnerability Spec rstack overflow:   Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] onnx==1.16.0
[pip3] onnxruntime-training==1.18.0
[pip3] optree==0.11.0
[pip3] pytorch-triton==3.0.0+dedb7bdf33
[pip3] torch==2.5.0.dev20240718+cu118
[pip3] torchaudio==2.4.0.dev20240718+cu118
[pip3] torchvision==0.20.0.dev20240718+cu118
[conda] Could not collect

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131194

Summary: Rather then using stdin/stdout for IPC, we can create new pipes and pass the descriptors to the subproc via the cmd line. https://github.com/pytorch/pytorch/issues/131070 reports an issue where the combination of deepspeed and onnxruntime-training causes _something_ in the subproc to write to stdout and corrupt the IPC. The current implementation was already brittle; we can just create new pipes specifically for the IPC.

Test Plan: I was able to repro the MemoryError in https://github.com/pytorch/pytorch/issues/131070 by installing deepspeed and onnxruntime-training. Verified this PR fixes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D59968362](https://our.internmc.facebook.com/intern/diff/D59968362)

### CODE REVIEW

File: torch/_inductor/compile_worker/subproc_pool.py [hunk 1]
```diff
@@ -21,20 +21,6 @@
 log = logging.getLogger(__name__)
 
 
-class Pipe(typing.Protocol):
```
Comment (masnesral) [inline torch/_inductor/compile_worker/subproc_pool.py:24]: I don't think this was really needed. typechecking gives me no errors.

**Comment by masnesral** [comment]
I followed @zdevito 's simple_worker_pool approach, forwarded by @eellison

**Review by malfet — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Review by atalman — APPROVED** [review]

**Comment by masnesral** [comment]
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59968362).

**Pull request closed without merging**